rust/arrow/src/compute/kernels/filter.rs

```diff
@@ -17,139 +17,466 @@
+        self.target_index += 1;
+    }
+
+    #[inline]
+    fn copy_null_bits(&mut self, source_index: usize, count: usize) {
+        for i in 0..count {
+            self.copy_null_bit(source_index + i);
         }
-        DataType::Timestamp(TimeUnit::Millisecond, _) => {
-            filter_array!(array, filter, TimestampMillisecondArray)
+    }
+
+    fn null_count(&self) -> usize {
+        self.null_count
+    }
+
+    fn null_buffer(&mut self) -> Buffer {
+        self.target_buffer.resize(self.target_index).unwrap();
+        // use mem::replace to detach self.target_buffer from self so that it can be returned
+        let target_buffer = mem::replace(&mut self.target_buffer, MutableBuffer::new(0));
+        target_buffer.freeze()
+    }
+}
+
+fn get_null_bit_setter<'a>(data_array: &'a impl Array) -> Box<CopyNullBit + 'a> {
+    if let Some(null_bitmap) = data_array.data_ref().null_bitmap() {
+        // only return an actual null bit copy implementation if null_bitmap is set
+        Box::new(NullBitSetter::new(null_bitmap))
+    } else {
+        // otherwise return a no-op copy null bit implementation
+        // for improved performance when the filtered array doesn't contain NULLs
+        Box::new(NullBitNoop::new())
+    }
+}
+
+// transmute filter array to u64
+// - optimize filtering with highly selective filters by skipping entire batches of 64 filter bits
+// - if the data array being filtered doesn't have a null bitmap, no time is wasted to copy a null bitmap
+fn filter_array_impl(
+    filter_context: &FilterContext,
+    data_array: &impl Array,
+    array_type: DataType,
+    value_size: usize,
+) -> Result<ArrayDataBuilder> {
+    if filter_context.filter_len > data_array.len() {
+        return Err(ArrowError::ComputeError(
+            "Filter array cannot be larger than data array".to_string(),
+        ));
+    }
+    let filtered_count = filter_context.filtered_count;
+    let filter_mask = &filter_context.filter_mask;
+    let filter_u64 = &filter_context.filter_u64;
+    let data_start = data_array.data_ref().buffers()[0].raw_data();
+    let mut value_buffer = MutableBuffer::new(filtered_count * value_size);
+    value_buffer.resize(filtered_count * value_size)?;
+    let mut value_position = value_buffer.raw_data_mut();
+    let mut null_bit_setter = get_null_bit_setter(data_array);
+    let null_bit_setter = null_bit_setter.as_mut();
+    let all_ones_batch = !0u64;
+
+    for (i, filter_batch) in filter_u64.iter().enumerate() {
+        // foreach u64 batch
+        let filter_batch = *filter_batch;
+        if filter_batch == 0 {
+            // if batch == 0: skip
+            continue;
+        } else if filter_batch == all_ones_batch {
+            // if batch == all 1s: copy all 64 values in one go
+            let data_index = i * 64;
+            let data_len = value_size * 64;
+            null_bit_setter.copy_null_bits(data_index, 64);
+            unsafe {
```

Comment:
The use of `unsafe` makes me nervous. If the performance improvement justifies using this, how can we be sure we have adequate testing for this?

Reply:
@andygrove  I have replaced the use of unsafe memcpy with copy_from_slice - this does reduce performance (by about 12%) but it is still much faster than the old implementation; here are the latest benchmark results:

filter u8 low selectivity                      time:   [130.74 us 131.37 us 132.19 us]
filter u8 high selectivity                     time:   [5.2031 us 5.2366 us 5.2764 us]
filter u8 very low selectivity                 time:   [12.353 us 12.542 us 12.759 us]

filter context u8 low selectivity              time:   [129.54 us 129.88 us 130.30 us]
filter context u8 high selectivity             time:   [1.7926 us 1.7974 us 1.8046 us]
filter context u8 very low selectivity         time:   [8.7700 us 8.7987 us 8.8342 us]

filter context u8 w NULLs low selectivity      time:   [150.36 us 151.09 us 152.01 us]
filter context u8 w NULLs high selectivity     time:   [2.4173 us 2.4882 us 2.5703 us]
filter context u8 w NULLs very low selectivity time:   [158.86 us 160.32 us 162.26 us]

filter context f32 low selectivity             time:   [123.38 us 124.34 us 126.18 us]
filter context f32 high selectivity            time:   [1.4836 us 1.4994 us 1.5297 us]
filter context f32 very low selectivity        time:   [19.422 us 19.653 us 19.932 us]

Reply:
Thanks @yordan-pavlov We can always consider the unsafe changes in a future PR, but removing them for now makes it much easier to get these changes merged.